text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,908 @@
+
+The main reason that this proposal does prefer to keep cancellation out of scope
+is to keep as close as possible to synchronous functions as possible. Those do
+not have a standardized cancellation intitiation mechansim either.
+
+A `CancellationToken/StopToken` can be standardized via a separate RFC, e.g.
+in a similiar form that C++ introduced via
+[`std::stop_token`](https://en.cppreference.com/w/cpp/thread/stop_token) and
+[P0660R9](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0660r9.pdf).
+
+A single standardized `stop_token` could cover the cancellation of asynchronous
+tasks as well as synchronous threads, and thereby also allow to gracefully
+shut down subsytems which make use of both.
+
+## Relation to `poll_drop`
+
+A proposal which might be related is the one about
+[`async drop / asynchronous destructors`](https://boats.gitlab.io/blog/post/poll-drop/).
+This blog post proposes to a `poll_drop` method to certain types,
+which allows to run asynchronous code also in their cleanup phase.
+However this mechanism would still not meet the motivations for this proposal.
+
+The reason for this is that **`poll_drop` code is not guaranteed to be run**.
+Since it is just an optional method on types, it might never be called.
+This is especially likely to happen if `poll_drop` is just used deep within the
+call hierarchy, while the code around it is not aware of it.
+
+Any code which is similiar to [`select!`](https://docs.rs/futures/0.3.4/futures/macro.select.html),
+and which was not converted to be made aware of `poll_drop`
+could cause `poll_drop` not be called. Therefore `poll_drop` is not a mechanism
+which e.g. allows us to add a safe Rust APIs around async completion based
+mechanism.
+
+`#[completion] async fn` in comparison enforces through the type system that
+methods will run to completion, and any necessary cleanup code is run before 
+resources are released.
+
+`poll_drop` and `#[completion] async fn` are also similar in the sense that they
+both require some `Future`s to be rewritten to support proper cleanup. The
+main difference here is that `#[completion] async fn` only requires new `Future`s
+that want to support running to completion to be rewritten, and will enforce
+this constraint through Rusts type system. `poll_drop` in comparison would require
+potentially every existing `Future` and `Future`-combinator to be changed, in
+order to provide a similary high guarantee of async cleanup code being performed.
+
+There exists also the possibility to introduce a combination of both proposals:
+
+Since `RunToCompletionFuture` is a new trait, it would be possible to add
+`fn poll_drop(self: Pin<&mut Self>)` on this trait, and enforce users to of the
+trait to use it: The `unsafe` contract in the trait could require callers to
+either `.poll()` the future to completion, or to switch over to calling `poll_drop`.
+
+However the benefit seems low. By doing this the Future would again move away
```

Comment:
Panic handling seems to require `poll_drop` or similar, unless one wants to force `panic=abort` for `async` code.

Reply:
I think I agree. The issue I find most hard is during parallelism, i.e. real threads. This occurs in multithreaded async runtimes, but also without async in regular threads. Crossbeam's scoped threads catch the panic and waits for the threads to terminate. Which can wait forever. It's somewhat terrifying that drop is unbounded in time, but similarly terrifying to not be able to statically borrow between threads.. I have no good answers here - just sharing my thoughts.

Reply:
Panic propagation is *already* unbounded in time, since we have Turing-complete destructors.  What this does mean is that we need `std::panic::catch_unwind` to be extremely fast, ideally free.